elfutils: Update to 0.192 and add a patch

Ra2-IFV opened the pull request:
build success on commit 5d7f4eaeced26a1c15c0d74d95086280be4f7aae

Ra2-IFV committed:
elfutils: Update to 0.192

Add a patch still under review to fix some errors.

Refresh patch:
- 003-libintl-compatibility.patch
- 005-build_only_libs.patch
- 006-Fix-build-on-aarch64-musl.patch

Add patch:
- 102-fix-potential-deref-of-null-error.patch

Release notes are in the link below.

Link: https://inbox.sourceware.org/elfutils-devel/CAJDtP-T3+gXqHWp3T0mejWWbPr0_1tHetEXwfB67-o+zz7ShiA@mail.gmail.com/T/#u
Signed-off-by: Ryan Keane <the.ra2.ifv@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16886
Signed-off-by: Robert Marko <robimarko@gmail.com>

Ra2-IFV committed:
elfutils: Backport some patches to fix errors

On aarch64 musl gcc 14.x compiler, trying compiling elfutils 0.192 with
lto option enabled will cause null-dereference error.
Example error message:

...
elf_compress.c: In function 'elf_compress':
elf_compress.c:675:26: error: potential null pointer dereference [-Werror=null-dereference]
  675 |           shdr->sh_flags |= SHF_COMPRESSED;
      |                          ^
elf_compress_gnu.c: In function 'elf_compress_gnu':
elf_compress_gnu.c:127:25: error: potential null pointer dereference [-Werror=null-dereference]
  127 |           shdr->sh_size = new_size;
      |                         ^                      ^
...

This is a false postive warning but will abort compilation if gcc has
`-Werror` flag. This commit add a patch for this, see the bugzilla
report below.

This commit backports a series of patches to fix some errors.

Add patch:
- 007-add-libeu-symbols-to-libelf.patch
- 008-fix-autoconf-ENABLE_IMA_VERIFICATION.patch
- 009-fix-null-dereference-with-lto.patch

Link: https://sourceware.org/bugzilla/show_bug.cgi?id=32311
Signed-off-by: Ryan Keane <the.ra2.ifv@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16886
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@mcprat You are the resident expert :)

Ra2-IFV commented:
Re-cloning repo just blows up my git config at repository level...

mcprat commented:
in this case the maintainer @luizluca should be considered the expert, not me haha

anyway,

in first commit I would include a patch that's still under review

https://patchwork.sourceware.org/project/elfutils/patch/20240820173716.13225-1-maks.mishinFZ@gmail.com/

in second commit I would include more backports

https://sourceware.org/git/?p=elfutils.git;a=commit;h=f5d6e088f84dd05278c4698a21cbf1ff4569978d
https://sourceware.org/git/?p=elfutils.git;a=commit;h=f3c664d069d81a4872a1ec8241ee709f37c53e9c


also, no need to put a link to upstream patches in the commit message
and the link to the bug report is better in tag format:

```
Link: https://sourceware.org/bugzilla/show_bug.cgi?id=32311
Signed-off-by: Ryan Keane <the.ra2.ifv@gmail.com>
```

mcprat commented:
then you can just name the second commit a variation of "...backport some fixes..."

luizluca commented:
> in this case the maintainer @luizluca should be considered the expert, not me haha

I'm the maintainer by accident, not experience ;-) Feel free to open a PR to take the role. You clearly are following elfutils much closer than me.

Ra2-IFV commented:
> in first commit I would include a patch that's still under review
> 
> https://patchwork.sourceware.org/project/elfutils/patch/20240820173716.13225-1-maks.mishinFZ@gmail.com/
> 
> in second commit I would include more backports
> 
> https://sourceware.org/git/?p=elfutils.git;a=commit;h=f5d6e088f84dd05278c4698a21cbf1ff4569978d
> https://sourceware.org/git/?p=elfutils.git;a=commit;h=f3c664d069d81a4872a1ec8241ee709f37c53e9c

I don't know why I should include these patches in my commit - I even don't know what do these patches come from, why I should include them and why only these patches.. Really new to here and don't know the workflow at all.. Should I close this PR and let others with greater abilities to do it?
For myself I was just having a problem with this package, reported in the upstream bug tracker and got the patch to fix it, so I open a PR for this fix and update the package version because there's a new release.

> also, no need to put a link to upstream patches in the commit message

I see, just make sure I follow the contributing guide.

> and the link to the bug report is better in tag format:
> 
> ```
> Link: https://sourceware.org/bugzilla/show_bug.cgi?id=32311
> Signed-off-by: Ryan Keane <the.ra2.ifv@gmail.com>
> ```

I learned that. Thanks for your advice.

mcprat commented:
> Feel free to open a PR to take the role. You clearly are following elfutils much closer than me.

I am following it closely at the moment, but I don't plan to for very long...

just because of the need to solve build problems on the host side, not because I know much about ELF...

mcprat commented:
> what do these patches come from, why I should include them and why only these patches

Those patches are just ones that look useful to have after I reviewed the proposed and accepted patches after the release.

The first patch has been pending for a long time, but is simple and looks fine

The second set of patches are from the maintainers of elfutils and are small issues they noticed after release, I refer to them as "backports" because they are accepted already but are not part of the new release. The patch you are adding already is also an example of a "backport".



> Should I close this PR and let others with greater abilities to do it?

having the ability to start a PR suggests you likely have the ability to edit it. It's not very difficult, and if you need help with git just ask us.

mcprat commented:
> > in first commit I would include a patch that's still under review
> > https://patchwork.sourceware.org/project/elfutils/patch/20240820173716.13225-1-maks.mishinFZ@gmail.com/

since patchwork is a pain to work with, here's the original email, for easier copy/pasting

https://sourceware.org/pipermail/elfutils-devel/2024q3/007341.html

Ra2-IFV commented:
> Those patches are just ones that look useful to have after I reviewed the proposed and accepted patches after the release.
> The first patch has been pending for a long time, but is simple and looks fine

I see, I'll take it

> The second set of patches are from the maintainers of elfutils and are small issues they noticed after release, I refer to them as "backports" because they are accepted already but are not part of the new release. The patch you are adding already is also an example of a "backport".

Just being a bit confused at here [Wiki: Naming patches](https://openwrt.org/docs/guide-developer/toolchain/use-patches-with-buildsystem#naming_patches)

> having the ability to start a PR suggests you likely have the ability to edit it. It's not very difficult, and if you need help with git just ask us.

:+1:

mcprat commented:
> Just being a bit confused at here [Wiki: Naming patches](https://openwrt.org/docs/guide-developer/toolchain/use-patches-with-buildsystem#naming_patches)

these notes are meant to describe linux kernel patches, for directories like
`target/linux/generic/pending-6.6`
because there are so many that it needs some organizational structure

here it is much more simple because there's only a few for each tool,
just 3 numbers with a basic meaning for each and summarize the title

`[category/type][series_number][patch_number]-fix-something-in-the-library.patch`

Ra2-IFV commented:
> > Just being a bit confused at here [Wiki: Naming patches](https://openwrt.org/docs/guide-developer/toolchain/use-patches-with-buildsystem#naming_patches)
> 
> these notes are meant to describe linux kernel patches, for directories like `target/linux/generic/pending-6.6` because there are so many that it needs some organizational structure
> 
> here it is much more simple because there's only a few for each tool, just 3 numbers with a basic meaning for each and summarize the title
> 
> `[category/type][series_number][patch_number]-fix-something-in-the-library.patch`

Very helpful! But patching for linux kernel sounds too scary to me, and I don't have the ability to do, yet
But I'll bookmark this! Thanks very much.

Ra2-IFV commented:
Build pass at commit d56c99a99cdadb5720027beb23332714b7a6f854 .
I have another PR at #15804 , for OpenSSL, also a critical package so it's still under reviews.

mcprat commented:
good job so far, almost there. now just some picky formality...

links should be at the bottom right before your signoff

title the second commit differently since it now does multiple backports for various  different fixes

it would be nice to order the patches according to the date they were made (see email headers in the first 4 lines of each patch), this decreases the probability of conflicts or unnecessary changes to the original copy of the patch when refreshing

Ra2-IFV commented:
Okay, not sure if this met your formality requirements, I am still learning ^^

mcprat commented:
looks pretty good

1. all tags at the bottom, signed-off-by is just another example of a tag
example of a long series of tags:

```
Link: ...
CC: ...
Fixes: ...
Acked-by: ...
Co-authored-by: ...
Signed-off-by: ...
Link: ...
Signed-off-by: ...
```


2. just delete this

```
Links to patch source:
- Link: https://sourceware.org/git/?p=elfutils.git;a=patch;h=8707194a9f2f0b13e53041b03ebfdbdbd2942e43
- Link: https://sourceware.org/git/?p=elfutils.git;a=patch;h=f5d6e088f84dd05278c4698a21cbf1ff4569978d
- Link: https://sourceware.org/git/?p=elfutils.git;a=patch;h=f3c664d069d81a4872a1ec8241ee709f37c53e9c
```

because each patch has the hash already on their first line,
it's not very useful to have the links as well, kind of like duplicate info

Ra2-IFV commented:
Sorry for being that dumb... It's really a bit difficult to find tutorials about `commit message tagging`
[OpenWrt Wiki - Submitting patches](https://openwrt.org/submitting-patches#sign_your_work)
[git - git-interpret-trailers - Add or parse structured information in commit messages](https://git-scm.com/docs/git-interpret-trailers)  
[Linux - Submitting patches: the essential guide to getting your code into the kernel](https://www.kernel.org/doc/Documentation/process/submitting-patches.rst)

mcprat commented on package/libs/elfutils/Makefile:
```diff
@@ -7,12 +7,13 @@
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
-PKG_SOURCE_URL:=https://sourceware.org/$(PKG_NAME)/ftp/$(PKG_VERSION)
-PKG_HASH:=df76db71366d1d708365fc7a6c60ca48398f14367eb2b8954efc8897147ad871
+PKG_SOURCE_URL:=https://sourceware.org/$(PKG_NAME)/ftp/$(PKG_VERSION) \
+								https://mirrors.kernel.org/sourceware/$(PKG_NAME)/$(PKG_VERSION)
+PKG_HASH:=616099beae24aba11f9b63d86ca6cc8d566d968b802391334c91df54eab416b4
```
there is a whitespace issue here, use 1 or 2 tabs and then spaces to align

also, I think adding a mirror should be in a separate commit,
either before or after the others, it doesn't matter where.

Ra2-IFV committed:
elfutils: Add mirrors.kernel.org as mirrror

Add mirrors.kernel.org as mirror, listed on sourceware mirror sites
page.

Link: https://sourceware.org/mirrors.html
Signed-off-by: Ryan Keane <the.ra2.ifv@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16886
Signed-off-by: Robert Marko <robimarko@gmail.com>

Ra2-IFV commented:
![IMG_20241114_035223_385](https://github.com/user-attachments/assets/6df20332-8eee-47aa-955d-7d5d34719ac9)

mcprat commented:
we're very close!

commit title has a typo

mcprat commented:
also for the last commit title you can say "backport some patches" instead of "add some patches" if you like, both are acceptable though

Ra2-IFV commented:
Oh no, a git rebase disaster

mcprat reviewed (approved):
looks good, I think it's ready @robimarko

robimarko commented:
Thanks! Rebased on top of main and merged!

mcprat commented:
side note to anyone who sees

for host tools, elfutils update is likely going to be handled as part of #16522
